File: solutions/rust/src/year2022/day02.rs
#[derive(Clone, Copy)]
enum RoundResult {
    Loss = 0,
    Draw = 3,
    Win = 6,
}

#[derive(Clone, Copy)]
enum Shape {
    Rock = 1,
    Paper = 2,
    Scissors = 3,
}

trait FromPlayerEncoding {
    fn from_player_encoding(encoded: &str) -> Self;
}

trait ComputeRoundPoints {
    fn compute_round_points(&self, oponent: &Shape) -> i32;
}

impl Shape {
    fn from_oponent_encodig(encoded: &str) -> Self {
        match encoded {
            "A" => Self::Rock,
            "B" => Self::Paper,
            "C" => Self::Scissors,
            _ => panic!("Unexpected oponent shape: '{encoded}'"),
        }
    }

    fn result_against(&self, oponent: &Self) -> RoundResult {
        use {RoundResult::*, Shape::*};

        match (self, oponent) {
            (Rock, Scissors) | (Paper, Rock) | (Scissors, Paper) => Win,

            (Scissors, Rock) | (Paper, Scissors) | (Rock, Paper) => Loss,

            _ => Draw,
        }
    }
}

impl RoundResult {
    fn necessary_shape_against(&self, oponent: &Shape) -> Shape {
        use {RoundResult::*, Shape::*};

        match (self, oponent) {
            (Loss, Paper) | (Draw, Rock) | (Win, Scissors) => Rock,

            (Loss, Scissors) | (Draw, Paper) | (Win, Rock) => Paper,

            _ => Scissors,
        }
    }
}

// Part 1
impl FromPlayerEncoding for Shape {
    fn from_player_encoding(encoded: &str) -> Self {
        match encoded {
            "X" => Self::Rock,
            "Y" => Self::Paper,
            "Z" => Self::Scissors,
            _ => panic!("Unexpected player shape: '{encoded}'"),
        }
    }
}

impl ComputeRoundPoints for Shape {
    fn compute_round_points(&self, oponent: &Shape) -> i32 {
        *self as i32 + self.result_against(oponent) as i32
    }
}

// Part 2
impl FromPlayerEncoding for RoundResult {
    fn from_player_encoding(encoded: &str) -> Self {
        match encoded {
            "X" => Self::Loss,
            "Y" => Self::Draw,
            "Z" => Self::Win,
            _ => panic!("Unexpected player result: '{encoded}'"),
        }
    }
}

impl ComputeRoundPoints for RoundResult {
    fn compute_round_points(&self, oponent: &Shape) -> i32 {
        *self as i32 + self.necessary_shape_against(oponent) as i32
    }
}

fn parse<P: FromPlayerEncoding>(input: &str) -> Vec<(Shape, P)> {
    let rounds = input.trim().lines().map(|line| {
        let mut split = line.split_whitespace();
        let oponent = Shape::from_oponent_encodig(split.next().unwrap());
        let player = P::from_player_encoding(split.next().unwrap());
        (oponent, player)
    });

    rounds.collect()
}

fn play_game(game: &Vec<(Shape, impl ComputeRoundPoints)>) -> i32 {
    game.iter()
        .map(|(oponent, player)| player.compute_round_points(oponent))
        .sum()
}

pub(crate) fn solution(input: &str) -> (String, String) {
    let part1: Vec<(Shape, Shape)> = parse(input);
    let part2: Vec<(Shape, RoundResult)> = parse(input);

    (play_game(&part1).to_string(), play_game(&part2).to_string())
}

File: solutions/zig/src/year2022/day02.zig
const std = @import("std");

const ParseError = error{
    InvalidOponentEncoding,
    InvalidPlayerEncoding,
};

const Shape = enum(usize) {
    rock = 1,
    paper = 2,
    scissors = 3,

    const Self = @This();

    pub fn fromOponentEncoding(encoded: u8) ParseError!Self {
        return switch (encoded) {
            'A' => .rock,
            'B' => .paper,
            'C' => .scissors,
            else => error.InvalidOponentEncoding,
        };
    }

    pub fn fromPlayerEncoding(encoded: u8) ParseError!Self {
        return switch (encoded) {
            'X' => .rock,
            'Y' => .paper,
            'Z' => .scissors,
            else => error.InvalidPlayerEncoding,
        };
    }

    pub fn resultAgainst(self: Self, oponent: Self) RoundResult {
        return switch (self) {
            .rock => switch (oponent) {
                .rock => .draw,
                .paper => .loss,
                .scissors => .win,
            },

            .paper => switch (oponent) {
                .rock => .win,
                .paper => .draw,
                .scissors => .loss,
            },

            .scissors => switch (oponent) {
                .rock => .loss,
                .paper => .win,
                .scissors => .draw,
            },
        };
    }

    pub fn calculateRoundPoints(self: Self, oponent: Shape) usize {
        return @intFromEnum(self) + @intFromEnum(self.resultAgainst(oponent));
    }
};

const RoundResult = enum(usize) {
    loss = 0,
    draw = 3,
    win = 6,

    const Self = @This();

    pub fn fromPlayerEncoding(encoded: u8) ParseError!Self {
        return switch (encoded) {
            'X' => .loss,
            'Y' => .draw,
            'Z' => .win,
            else => error.InvalidPlayerEncoding,
        };
    }

    pub fn necessaryShapeAgainst(self: Self, oponent: Shape) Shape {
        return switch (self) {
            .loss => switch (oponent) {
                .rock => .scissors,
                .paper => .rock,
                .scissors => .paper,
            },

            .win => switch (oponent) {
                .rock => .paper,
                .paper => .scissors,
                .scissors => .rock,
            },

            .draw => oponent,
        };
    }

    pub fn calculateRoundPoints(self: Self, oponent: Shape) usize {
        return @intFromEnum(self) + @intFromEnum(self.necessaryShapeAgainst(oponent));
    }
};

fn Round(comptime T: type) type {
    return struct {
        oponent: Shape,
        player: T,

        const Self = @This();

        pub fn calculatePoints(self: Self) usize {
            return self.player.calculateRoundPoints(self.oponent);
        }
    };
}

fn parse(comptime T: type, input: []const u8, allocator: std.mem.Allocator) !std.ArrayList(Round(T)) {
    var game = std.ArrayList(Round(T)).init(allocator);
    var lines = std.mem.split(u8, std.mem.trim(u8, input, " \n"), "\n");
    while (lines.next()) |line| {
        const oponent = try Shape.fromOponentEncoding(line[0]);
        const player = try T.fromPlayerEncoding(line[2]);
        try game.append(.{ .oponent = oponent, .player = player });
    }

    return game;
}

fn playGame(comptime T: type, game: *const std.ArrayList(Round(T))) usize {
    var sum: usize = 0;
    for (game.items) |round| {
        sum += round.calculatePoints();
    }

    return sum;
}

pub fn solution(input: []const u8, allocator: std.mem.Allocator) !struct { []u8, []u8 } {
    const part1 = try parse(Shape, input, allocator);
    defer part1.deinit();

    const part2 = try parse(RoundResult, input, allocator);
    defer part2.deinit();

    return .{
        try std.fmt.allocPrint(allocator, "{}", .{playGame(Shape, &part1)}),
        try std.fmt.allocPrint(allocator, "{}", .{playGame(RoundResult, &part2)}),
    };
}
